db/migrate/20181201000100_create_message_recipients.foodsoft_messages.rb

```diff
@@ -0,0 +1,44 @@
+class CreateMessageRecipients < ActiveRecord::Migration
+  def up
+    create_table :message_recipients do |t|
+      t.references :message, index: true, null: false
+      t.references :user, null: false
+      t.integer :email_state, default: 0, null: false
+      t.datetime :read_at
+    end
+
+    add_index :message_recipients, [:user_id, :read_at]
+
+    Message.all.each do |m|
+      recipients = YAML.load(m.recipients_ids).map do |r|
+        {
+          message_id: m.id,
+          user_id: r,
+          email_state: :sent
+        }
+      end
```

Comment:
When referencing models from a migration, it is important to define them first. In a future version of Foodsoft this model may have been removed, even though it is still used in this migration. So inside the class, at the top, put `class Message < ActiveRecord::Base; end`. Since we only use database fields here, that will do.

Reply:
our migration is already broken in that direction. IMHO we should only support migration from one release to next release.
e.g.: I'd like to change our wiki engine in the future. for the migrations i'll need the old engine to parse the pages, but it don't want to keep it until the end of time as dependency. i'd jut add an error message into the migration, when we remove the dependency, directing the user to migrate to a previous version first

Reply:
Hmm that is a good point, about the wiki.
Still I'd rather keep migration compatibility where possible. I agree that we can support migration from one major version to another, but it would still be nice if it would work across multiple versions for most of the time (and otherwise certainly noted in the release notes).

Reply:
There is a dependency on `message_recipients` in line 36, so a simple `ActiveRecord::Base` is not enough

Reply:
Then it would be
```ruby
Class Message < ActiveRecord::Base
  has_many :message_recipients
  has_many :recipients, through: :message_recipients, source: :user
end
```

Reply:
makes sense 😄